File: knhmp/src/main/java/io/github/fopwoc/knhmp/KnhMpSourceSets.java
package io.github.fopwoc.knhmp;

import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.gradle.api.Action;

public final class KnhMpSourceSets {

    private final Map<String, KnhMpSourceSet> sourceSets = new LinkedHashMap<>();

    public KnhMpSourceSet sourceSet(String name) {
        return sourceSets.computeIfAbsent(name, KnhMpSourceSet::new);
    }

    public void sourceSet(String name, Action<? super KnhMpSourceSet> configure) {
        configure.execute(sourceSet(name));
    }

    public void commonMain(Action<? super KnhMpSourceSet> configure) {
        configure.execute(getCommonMain());
    }

    public void gtnhMain(Action<? super KnhMpSourceSet> configure) {
        configure.execute(getGtnhMain());
    }

    public void fabricLegacyMain(Action<? super KnhMpSourceSet> configure) {
        configure.execute(getFabricLegacyMain());
    }

    public void fabricMain(Action<? super KnhMpSourceSet> configure) {
        configure.execute(getFabricMain());
    }

    public void neoforgeMain(Action<? super KnhMpSourceSet> configure) {
        configure.execute(getNeoforgeMain());
    }

    public KnhMpSourceSet getCommonMain() {
        return sourceSet("commonMain");
    }

    public KnhMpSourceSet getGtnhMain() {
        return sourceSet("gtnhMain");
    }

    public KnhMpSourceSet getFabricLegacyMain() {
        return sourceSet("fabricLegacyMain");
    }

    public KnhMpSourceSet getFabricMain() {
        return sourceSet("fabricMain");
    }

    public KnhMpSourceSet getNeoforgeMain() {
        return sourceSet("neoforgeMain");
    }

    List<String> closure(String leaf) {
        LinkedHashSet<String> result = new LinkedHashSet<>();
        collect(leaf, result, new LinkedHashSet<>());
        return new ArrayList<>(result);
    }

    int effectiveJvmTarget(String leaf, int backendMinimum) {
        int sourceRequirement = closure(leaf).stream()
                .map(sourceSets::get)
                .map(KnhMpSourceSet::getJvmTarget)
                .filter(target -> target != null)
                .mapToInt(Integer::intValue)
                .max()
                .orElse(backendMinimum);
        return Math.max(sourceRequirement, backendMinimum);
    }

    Set<String> names() {
        return Set.copyOf(sourceSets.keySet());
    }

    Collection<KnhMpSourceSet> all() {
        return List.copyOf(sourceSets.values());
    }

    boolean isEmpty() {
        return sourceSets.isEmpty();
    }

    private void collect(String name, LinkedHashSet<String> result, LinkedHashSet<String> visiting) {
        KnhMpSourceSet sourceSet = sourceSets.get(name);
        if (sourceSet == null) {
            throw new IllegalStateException("Unknown source set: " + name);
        }
        if (!visiting.add(name)) {
            throw new IllegalStateException("Source-set cycle: " + visiting + " -> " + name);
        }
        for (String parent : sourceSet.parents()) {
            collect(parent, result, visiting);
        }
        visiting.remove(name);
        result.add(name);
    }
}

File: knhmp/src/main/kotlin/io/github/fopwoc/knhmp/KnhMpSourceSets.kt
package io.github.fopwoc.knhmp

import org.gradle.api.Action

class KnhMpSourceSets {

    private val sourceSets = linkedMapOf<String, KnhMpSourceSet>()

    val commonMain: KnhMpSourceSet get() = sourceSet("commonMain")
    val gtnhMain: KnhMpSourceSet get() = sourceSet("gtnhMain")
    val fabricLegacyMain: KnhMpSourceSet get() = sourceSet("fabricLegacyMain")
    val fabricMain: KnhMpSourceSet get() = sourceSet("fabricMain")
    val neoforgeMain: KnhMpSourceSet get() = sourceSet("neoforgeMain")

    fun sourceSet(name: String): KnhMpSourceSet =
        sourceSets.getOrPut(name) { KnhMpSourceSet(name) }

    fun sourceSet(name: String, configure: Action<in KnhMpSourceSet>) =
        configure.execute(sourceSet(name))

    fun commonMain(configure: Action<in KnhMpSourceSet>) = configure.execute(commonMain)

    fun gtnhMain(configure: Action<in KnhMpSourceSet>) = configure.execute(gtnhMain)

    fun fabricLegacyMain(configure: Action<in KnhMpSourceSet>) = configure.execute(fabricLegacyMain)

    fun fabricMain(configure: Action<in KnhMpSourceSet>) = configure.execute(fabricMain)

    fun neoforgeMain(configure: Action<in KnhMpSourceSet>) = configure.execute(neoforgeMain)

    internal fun closure(leaf: String): List<String> = buildList {
        collect(leaf, this, linkedSetOf())
    }

    internal fun effectiveJvmTarget(leaf: String, backendMinimum: Int): Int =
        closure(leaf)
            .mapNotNull { sourceSets.getValue(it).jvmTarget }
            .maxOrNull()
            ?.coerceAtLeast(backendMinimum)
            ?: backendMinimum

    internal fun names(): Set<String> = sourceSets.keys.toSet()

    internal fun all(): Collection<KnhMpSourceSet> = sourceSets.values.toList()

    internal val isEmpty: Boolean get() = sourceSets.isEmpty()

    private fun collect(name: String, result: MutableList<String>, visiting: MutableSet<String>) {
        val sourceSet = checkNotNull(sourceSets[name]) { "Unknown source set: $name" }
        check(visiting.add(name)) { "Source-set cycle: $visiting -> $name" }
        sourceSet.parents().forEach { collect(it, result, visiting) }
        visiting.remove(name)
        if (name !in result) result += name
    }
}
